Add ECDSA support (named curve only, requires OpenSSL 1.1.0)

tatsuhiro-t opened the pull request:
This commit adds ECDSA support.  Currently, only named curve is
supported.  OpenSSL 1.1.0, which is not released yet, added necessary
functions to support EC_KEY with custom engine, and prior versions
have no workaround over these missing functions, we have to depend on
OpenSSL 1.1.0.  More specifically, we require OpenSSL 1.1.0 pre2.

tatsuhiro-t committed:
Add ECDSA support (named curve only, requires OpenSSL 1.1.0)

This commit adds ECDSA support.  Currently, only named curve is
supported.  OpenSSL 1.1.0, which is not released yet, added necessary
functions to support EC_KEY with custom engine, and prior versions
have no workaround over these missing functions, we have to depend on
OpenSSL 1.1.0.  More specifically, we require OpenSSL 1.1.0 pre2.

tatsuhiro-t committed:
Compile with openssl 1.1.0pre5

tatsuhiro-t commented:
I've added new commit to compile with OpenSSL 1.1.0pre5.  It contains massive changes, and it broke RSA code in neverbleed.   1.1.0 makes many OpenSSL object struct opaque, and all codes to access field directly break.  This commit fixes this issue.

tatsuhiro-t commented:
Since OpenSSL 1.1.0 is around the corner (the OpenSSL dev said that they will release it in August 25, 2016), it is a good time for us to seriously consider to review this PR.

tatsuhiro-t committed:
Merge branch 'master' into ecdsa

tatsuhiro-t commented:
Merged master branch, resolved conflict.

tatsuhiro-t committed:
Fix compiler warning with openssl 1.1.0

tatsuhiro-t committed:
Only consider openssl 1.1.0 final, and introduce macro to simplify code

tatsuhiro-t commented:
Now we only consider openssl 1.1.0 final version.  I also fixed compiler warnings.

kazuho commented on neverbleed.c:
```diff
@@ -568,23 +778,47 @@ int neverbleed_load_private_key_file(neverbleed_t *nb, SSL_CTX *ctx, const char
+        break;
+#endif
+    default:
+        if ((errstr = expbuf_shift_str(&buf)) == NULL) {
```
I assume we should just return an error to the peer without trying to parse arguments calling by expbuf_shift_xxx, considering the fact that we cannot determine the number and types of the arguments unless the type is TYPE_RSA or TYPE_ECDSA.

tatsuhiro-t replied:
I'm ok with that.  But we are sure that after key_index, error message string follows.  Can we just use it to return the error message to the client?

kazuho commented on neverbleed.c:
```diff
@@ -45,6 +45,10 @@
 #endif
 #include "neverbleed.h"
 
+#define OPENSSL_101_API (!defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x1010000fL)
```
How about renaming this to `OPENSSL_1_1_API`. IMO, `101` is confusing with `1.0.1`.

tatsuhiro-t replied:
Agreed.

tatsuhiro-t replied:
Fixed

kazuho commented on neverbleed.c:
```diff
@@ -568,23 +778,47 @@ int neverbleed_load_private_key_file(neverbleed_t *nb, SSL_CTX *ctx, const char
+        pkey = create_pkey(nb, key_index, estr, nstr);
+        break;
+#if OPENSSL_101_API
+    case NEVERBLEED_TYPE_ECDSA:
```
It _might_ be a good idea to create a block within the case (i.e. `case NEVERBLEED_TYPE_ECDSA: { ... } break;`) and declare the variables only for ECDSA-related operation to suppress following warnings:

```
neverbleed.c:770:11: warning: unused variable 'ec_pubkeystr' [-Wunused-variable]
    char *ec_pubkeystr;
          ^
neverbleed.c:771:12: warning: unused variable 'curve_name' [-Wunused-variable]
    size_t curve_name;
           ^
```

And the same thing could be done for RSA as well, I guess.

tatsuhiro-t replied:
Good idea. Will do.

tatsuhiro-t replied:
Fixed

kazuho commented on neverbleed.c:
```diff
@@ -596,6 +830,14 @@ static int load_key_stub(struct expbuf_t *buf)
     char *estr = NULL, *nstr = NULL, errbuf[NEVERBLEED_ERRBUF_SIZE] = "";
+    size_t type = NEVERBLEED_TYPE_NONE;
+    EC_KEY *ec_key = NULL;
+    EVP_PKEY *pkey;
```
I see the following warnings when trying to compile the file with OpenSSL 1.0.2.

I assume that use of `pkey` without initialization in an error. Also, it might be preferable to suppress the unused warnings as disscussed on https://github.com/h2o/neverbleed/pull/11/files#r76543935.

```
neverbleed.c:837:21: warning: unused variable 'ec_pubkey' [-Wunused-variable]
    const EC_POINT *ec_pubkey;
                    ^
neverbleed.c:834:13: warning: unused variable 'ec_key' [-Wunused-variable]
    EC_KEY *ec_key = NULL;
            ^
neverbleed.c:847:9: warning: variable 'pkey' is used uninitialized whenever 'if' condition is true
      [-Wsometimes-uninitialized]
    if ((fp = fopen(fn, "rt")) == NULL) {
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
neverbleed.c:909:9: note: uninitialized use occurs here
    if (pkey != NULL)
        ^~~~
neverbleed.c:847:5: note: remove the 'if' if its condition is always false
    if ((fp = fopen(fn, "rt")) == NULL) {
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
neverbleed.c:835:19: note: initialize the variable 'pkey' to silence this warning
    EVP_PKEY *pkey;
                  ^
                   = NULL
```

tatsuhiro-t replied:
Fixed

kazuho commented:
Thank you for the updates. I like how the RSA-related APIs are backported? for 1.0.2.

LGTM aside from my comments i-line (including the issues I found when trying to compile the file using OpenSSL 1.0.2).

tatsuhiro-t committed:
Rename OPENSSL_101_API as OPENSSL_1_1_API

tatsuhiro-t committed:
Fix warnings with -Wunused-variable and -Wsometimes-uninitialized

Reduce scope of some variables which are only used for RSA or ECDSA so
that they are defined where they are used.

kazuho commented on neverbleed.c:
```diff
@@ -606,27 +852,81 @@ static int load_key_stub(struct expbuf_t *buf)
     expbuf_dispose(buf);
-    expbuf_push_num(buf, rsa != NULL);
+    expbuf_push_num(buf, type != NEVERBLEED_TYPE_NONE);
+    expbuf_push_num(buf, type);
```
Would the code become more concise if `NEVERBLEED_TYPE_NONE` could be used as an indication of an error, instead of sending a boolean value indicating success _and_ the type value?
(and if we are to use the _none_ type as error indication it might make more sense to rename it the _error_ type)

tatsuhiro-t replied:
If I understand it correctly, you are suggesting that remove boolean (type != NEVERBLEED_TYPE_NONE), and rename neverbleed_type as neverbleed_error_type.  Is that correct?

tatsuhiro-t replied:
Perhaps, renaming NEVERBLEED_TYPE_NONE as NEVERBLEED_TYPE_ERROR ?

kazuho replied:
Correct.

At the moment, the arguments used for `load_key` are: ret, type, key_index, …

However, considering the fact that ret is a boolean that just indicates either success or failure and that we have `NEVERBLEED_TYPE_NONE`, it seems to me that the argument should better be: type, key_index, ..., merging the type and ret into a single argument.

So to rephrase my suggestion is:
- change `NEVERBLEED_TYPE_NONE` to `NEVERBLEED_TYPE_ERROR`
- stop passing `ret` as the first value of `load_key`, but use `type != NEVERBLEED_TYPE_ERROR` to see if the call has succeeded

tatsuhiro-t replied:
Thank you for clarification.  Will do.

tatsuhiro-t replied:
Done

kazuho commented:
Thank you for the changes.

The code looks great, however I left a comment since we might be possible to make it even better.

tatsuhiro-t committed:
Determine failure by type

We renamed NEVERBLEED_TYPE_NONE as NEVERBLEED_TYPE_ERROR, and instead
of returning boolean value to indicate success/failure, we use type
for its purpose.  Now boolean value was removed.

kazuho commented:
Thank you for the update! Merged to master.
